geps/gep-3155/index.md

```diff
@@ -0,0 +1,264 @@
+
+Configuring client certificate on the service level may result in multiple Gateways
+from different vendors sharing the same identity when connecting to a single service.
+
+However, given that:
+
+1. per-service certificate configuration is a niche configuration, required to support
+application-specific compatbility corner-cases
+1. cross-namespace service usage is protected by the reference grants
+
+this limitations should consitute an acceptable trade off.
+
+The possible need for more fine-grained settings can be mitigated by allowing shared 
+policies to selectively target specific Gateway instances, GatewayClasses or gateway 
+namespaces.
+
+### SANs on BackendTLSPolicy
+
+This change enables the certificate to have a different identity than the SNI
+(both are currently tied to the hostname field). This is particularly useful
+when using SPIFFE, which relies on URI Subject Names which are not valid SNIs 
+as per https://www.rfc-editor.org/rfc/rfc6066.html#section-3.
+
+In such case either connection properties or an arbitrary SNI, like cluster-local 
+service name could be used for certificate selection, while the identity validation
+will be done based on SubjectAltNames field.
+
+**1. Add a new `SubjectAltNames` field to `BackendTLSPolicyValidation`**
+
+```go
+type BackendTLSPolicyValidation struct {
+  // SubjectAltNames contains one or more Subject Alternative Names.
+  // When specified, at least one of certificate's Subject Alternate Names MUST 
+  // match at least one of the specified SubjectAltNames.
+  // +kubebuilder:validation:MaxItems=5
+  SubjectAltNames []SubjectAltName `json:"subjectAltNames,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=Cookie;Header
+type SubjectAltNameType string
+
+const (
+   // HostnameSubjectAltNameType specifies hostname-based SAN.
+   //
+   // Support: Core
+   HostnameSubjectAltNameType SubjectAltNameType = "Hostname"
+
+   // URISubjectAltNameType specifies URI-based SAN, e.g. SPIFFE id.
+   //
+   // Support: Core
+   URISubjectAltNameType SubjectAltNameType = "URI"
+)
+
+
+type SubjectAltName struct {
+  // Type determines the format of the Subject Alternative Name. Always required.
+  Type SubjectAltNameType `json:"type"`
+
+  // Hostname contains Subject Alternative Name specified in DNS name format. Required when Type is set to Hostname, ignored otherwise.
+  Hostname v1.PreciseHostname `json:"hostname,omitempty"`
+
+  // URI contains Subject Alternative Name specified in URI format. Required when Type is set to URI, ignored otherwise.
+  URI string `json:"uri,omitempty"`
+}
+```
+
+**2. Modify Spec for `BackendTLSPolicyValidation` `Hostname`**
+
+Before:
+```go
+  // 2. Hostname MUST be used for authentication and MUST match the certificate
+  //    served by the matching backend.
+```
+
+After:
+```go
+  // 2. Only if SubjectAltNames is not specified, Hostname MUST be used for 
+  //    authentication and MUST match the certificate served by the matching
+  //    backend.
```

Comment:
So, `Hostname` still sets the SNI that is sent as part of the initial request, but now no longer sets anything about how the server certificate must be checked?

In that case, I'd recommend something like this:
```suggestion
  // 2. Hostname MUST be used for authentication and MUST match the certificate
  //     served by the matching backend UNLESS SubjectAltNames is specified. If
  //     SubjectAltNames is specified, Hostname MUST only be used for setting
  //     the SNI sent, and the SubjectAltNames MUST be used for validating the
  //     serving certificate.
```

It's saying the same thing, but it is leaving less room for misunderstanding.

Reply:
Thanks a lot!